internal/migration/migration.go

```diff
@@ -0,0 +1,478 @@
+	}
+
+	if err := schema.Migrate(ctx, conn); err != nil {
+		err := fmt.Errorf("migrating error: %w", err)
+		return nil, nil, err
+	}
+
+	store, err := pgdb.NewStore(ctx, logger, conn, beaconName)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	cancel := func() {
+		_ = store.Close(ctx)
+		_ = conn.Close()
+	}
+
+	return store, cancel, err
+}
+
+func shouldMigratePostgres(ctx context.Context, logger log.Logger, beaconName, pgDSN string) error {
+	store, cancel, err := pgConn(ctx, logger, beaconName, pgDSN)
+	if err != nil {
+		return err
+	}
+	defer cancel()
+
+	storeLen, err := store.Len(ctx)
+	if err != nil {
+		return err
+	}
+
+	if storeLen != 0 {
+		logger.Warnw("Postgres contains beacon rounds. Skipping migration", "rounds count", storeLen)
+		return ErrMigrationNotNeeded
+	}
+
+	return nil
+}
+
+func (m *migrator) doMigrate(ctx context.Context) error {
+	defer func() {
+		finishedIn := time.Since(m.startedAt).String()
+		m.logger.Infow(
+			"Finished processing beacons",
+			"beaconName", m.beaconName,
+			"finishedIn", finishedIn,
+		)
+	}()
+
+	//nolint:exhaustive // We want to explicitly ignore the chain.MemDB backend since there's nothing to migrate there.
+	switch m.destination {
+	case chain.BoltDB:
+		go m.migrateBolt()
+	case chain.PostgreSQL:
+		go m.migratePostgres(ctx)
+	}
+
+	// Spin up the reader
+	go m.reader()
+
+	return m.wait()
+}
+
+//nolint:funlen // This function has the right length
+func (m *migrator) migratePostgres(ctx context.Context) {
+	rows := 0
+	defer func() {
+		if m.existingRows != rows {
+			err := fmt.Errorf("not all rounds migrated successfully expected: %d actual: %d", m.existingRows, rows)
+			m.errChan <- err
+			return
+		}
+	}()
+	defer func() {
+		finishedIn := time.Since(m.startedAt).String()
+		m.logger.Infow("finished saving data in postgres", "rows", rows, "finishedIn", finishedIn)
+	}()
+	defer m.ctxCancel()
+	defer func() {
+		if m.existingRows != rows {
+			m.logger.Errorw("not all rounds migrated successfully", "expected", m.existingRows, "actually", rows)
+			return
+		}
+	}()
+
+	store, cancel, err := pgConn(ctx, m.logger, m.beaconName, m.pgDSN)
+	if err != nil {
+		m.errChan <- err
+		return
+	}
+	defer cancel()
+
+	buffSize := 0
+
+	m.logger.Infow("dropping the FK from the table")
+	err = store.DropFK(ctx)
+	if err != nil {
+		m.errChan <- err
+		return
+	}
+	defer func() {
+		m.logger.Infow("adding FK back to the table")
+		err := store.AddFK(ctx)
+		if err != nil {
+			m.errChan <- err
+			return
+		}
+	}()
+
+	// Make sure we can still commit to the database
+	// avoiding the error:
+	//   pq: got 9229389 parameters but PostgreSQL only supports 65535 parameters
+	pgBuffSize := m.bufferSize
+	//nolint:gomnd // See below
+	if pgBuffSize > 30_000 {
+		//nolint:gomnd // See below
+		pgBuffSize = 30_000
+		m.logger.Warnw("buffer size automatically reconfigured for Postgres only", "bufferSize", pgBuffSize)
+	}
+
+	bs := make([]chain.Beacon, pgBuffSize)
+
+	for val := range m.distChan {
```

Comment:
Move the declaration of the rows variable above this for loop

Reply:
I can't, it's used in the defer at the top of the function.